### PR TITLE
Query planner optimisation: combining node and edge fragment cost, improving greedy traversal

### PR DESCRIPTION
**Pull request by duckofyork**

### CODE REVIEW

**Comment by duckofyork** [comment]
!rtg

**Comment by duckofyork** [comment]
!rtg

**Comment by duckofyork** [comment]
!rtg

File: grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java [hunk 1]
```diff
@@ -257,6 +258,27 @@ private static void greedyTraversal(List<Fragment> plan, Arborescence<Node> arbo
...
+        }
+        return weight[0];
+    }
+
+    private static double nodeFragmentWeight(Node node) {
+        return node.getFragmentsWithoutDependency().stream()
+                .map(Fragment::fragmentCost).reduce(0D, (x, y) -> x + y) +
+                node.getFragmentsWithDependencyVisited().stream()
+                        .map(Fragment::fragmentCost).reduce(0D, (x, y) -> x + y) +
+                node.getFixedFragmentCost();
```
Comment (aelred) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java]: Streams have a method `count()` that does what you want here. Add some variables as well to make it a bit clearer :P

Comment (duckofyork) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java]: I guess you mean method `sum()`, as fragment cost is not 1

Comment (aelred) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java]: Oops! You're right. There is still a way to do it even neater:
`.mapToDouble(Fragment::fragmentCost).sum();`

Comment (duckofyork) [reply grakn-graql/src/main/java/ai/grakn/graql/internal/gremlin/GreedyTraversalPlan.java]: I think `sum()` only supports integer stream.

**Comment by duckofyork** [comment]
@aelred merge merge merge!

**Pull request merged**